lib/ecto/repo/schema.ex

```diff
@@ -96,11 +99,42 @@ defmodule Ecto.Repo.Schema do
 
     if has_query? do
       rows = plan_query_in_rows(rows, header, adapter)
-      {rows, header, placeholder_vals_list}
+      {rows, header, placeholder_vals_list, counter}
     else
-      {rows, header, placeholder_vals_list}
+      {rows, header, placeholder_vals_list, counter}
     end
   end
+  defp extract_header_and_fields(%Ecto.Query{select: %Ecto.Query.SelectExpr{expr: {:%{}, _ctx, args}}} = query, _schema, _dumper, _autogen_id, _placeholder_map, adapter) do
+    header = Enum.map(args, &elem(&1, 0))
+
+    {query, params} = Ecto.Adapter.Queryable.plan_query(:all, adapter, query)
```

Comment:
I wonder if this should be `:insert_all`. Looking at ecto/repo/queryable.ex, we may also want to call both

    {query, opts} = repo.prepare_query(operation, query, opts)
    query = attach_prefix(query, opts)

(perhaps we should add these two calls inside `Ecto.Adapter.Queryable.plan_query`).

Reply:
and then call attach_prefix only if `operation == :insert_all`? Also, how do we get the `repo` in there?

Reply:
Oh, I was looking at the wrong file

Reply:
Good point. If we can't get the repo there we will need to do this dance before calling plan_adapter. Note you will have to call `Ecto.Queryable.to_query(queryable)` before calling prepare_query, so you might as well skip calling plan_adapter altogether and have all the logic here.

Reply:
There is another approach to take here, which is to call `Planner.query(query, operation, cache, adapter, 0)`. The advantage of calling it is that queries can be prepared and cached but that comes with other difficulties in the case of insert (since the cache key needs to include options such as conflict_target and so on - which is probably better to skip).

Reply:
Since attach_prefix was a `defp` in Ecto.Repo.Queryable and I need it in Ecto.Repo.Schema, I extracted the function into a new module Ecto.Repo.Common

Reply:
Please move it to Ecto.Query.Planner, as it has the shared functionality already. :)

Reply:
I've addressed this now. I have read the postgres insert into spec and there was no mention of limited functionality for the source query. So I've copied all the behaviour specific to `:all` queries in Ecto.Query.Planner to the new `:insert_all` operation.

Reply:
I was able to loop the repo from `Ecto.Repo.Schema.insert_all` through `do_insert_all` and finally into `extract_header_and_fields`. Only question is: do we prepare the query before we plan or the other way around?